tower-http/src/services/fs/serve_dir.rs

```diff
@@ -406,58 +456,118 @@ impl Future for ResponseFuture {
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         match &mut self.inner {
-            Inner::Valid(open_file_future) => match ready!(Pin::new(open_file_future).poll(cx)) {
-                Ok(Output::File(file, mime, chunk_size, maybe_encoding)) => {
-                    let body = AsyncReadBody::with_capacity(file, chunk_size).boxed();
-                    let body = ResponseBody::new(body);
-
-                    let mut res = Response::new(body);
-                    res.headers_mut().insert(header::CONTENT_TYPE, mime);
-
-                    if let Some(encoding) = maybe_encoding {
-                        res.headers_mut()
-                            .insert(header::CONTENT_ENCODING, encoding.into_header_value());
+            Inner::Valid(open_file_future) => {
+                let response = match ready!(Pin::new(open_file_future).poll(cx)) {
+                    Ok(Output::File(file_request)) => {
+                        let (maybe_file, size) = match file_request.extent {
+                            FileRequestExtent::Full(file, meta) => (Some(file), meta.len()),
+                            FileRequestExtent::Head(meta) => (None, meta.len()),
+                        };
+                        let mut builder = Response::builder()
+                            .header(header::CONTENT_TYPE, file_request.mime_header_value)
+                            .header(header::ACCEPT_RANGES, "bytes")
+                            .header(header::CONTENT_LENGTH, size.to_string());
+                        if let Some(encoding) = file_request.maybe_encoding {
+                            builder = builder
+                                .header(header::CONTENT_ENCODING, encoding.into_header_value());
+                        }
+                        if let Some(reasonable_range) = file_request.maybe_range {
+                            match reasonable_range {
+                                Ok(ranges) => {
```

Comment:
To avoid some of the extra layers of the extra conditionals it might be worth matching on `Ok(Some(range))` and `Ok(None)`